### ISSUE TITLE
TLS negociation fails when using a session stored in Session_Manager_In_Memory

### ISSUE THREAD

**Issue by louiz** [randombit/botan#1276]
* Connect to a remote server (I can reproduce 100% of the time on chat.freenode.net:6697 and google.com:443 for example)
* Negociate TLS from scratch, until the session is established (`Botan::TLS::Callbacks::tls_session_established()` is called), and store that into some Session_Manager_In_Memory object (by just passing that object to the Botan::TLS::Client constructor)
* Close the connexion
* Re-connect to the exact same server, same hostname etc
* Try to re-establish TLS, using the same Session_Manager_In_Memory object
* Fail with “Alert: decode_error”

There’s no problem if you just insert a “session_manager.remove_all()” before the second attempt.

Here is a short (and a little messy, sorry about that) code that reproduce the issue 100% of the time on my machine with the latest git botan on linux (fedora 26): https://gist.github.com/louiz/87b5eb4785eee1e31ff8e3cb7bed5d4a

compile with `g++ -std=c++11 main.cpp -I/usr/include/botan-2 -lbotan-2`

Example output:

```tls_emit_data: 210 bytes
received:1448
received:2896
Checking certificate, but actually just accepting it anyway.
tls_emit_data: 12 bytes
tls_emit_data: 267 bytes
tls_emit_data: 6 bytes
tls_emit_data: 45 bytes
received:342
tls_session_established
received:242
CLOSING, RECONNECTING.
tls_emit_data: 409 bytes
Alert: decode_error
received:7
Connection closed by remote        
```

(session_manager.remove_all() can be added between the two connexions, around line 131, to work-around the issue)

I didn’t dive into the code yet to find a clue of where the problem could lie.

**Comment by randombit** [comment]
Ugh. Confirmed.

Easy repro if sqlite3 is enabled:

$ ./botan tls_client gmail.com --session-db=sessions.db
// connects normally
./botan tls_client gmail.com --session-db=sessions.db
Alert: decode_error

**Comment by randombit** [comment]
Interesting, if Botan is the server no problem.

OpenSSL `s_server` says 
139822343451712:error:1417D0E3:SSL routines:tls_process_client_hello:parse tlsext:ssl/statem/statem_srvr.c:1209:

So probably the problem is we are sending an extension we should not send in a resumption.

**Comment by randombit** [comment]
This seems to be due to sending certificate status request (OCSP stapling request) on a resumption client hello. And admittedly that is a silly thing to ask for since no certificate is sent by the server in that situation.

**Comment by randombit** [comment]
Should be fixed now with d2630df. Thanks very much for the report.

For reasons I don't understand yet, resumption attempts don't actually work for some hosts including chat.freenode.net:6697 - this is #1277. The connection will succeed but each time a new session is created.

**Comment by louiz** [comment]
Thank you very much for the quick fix :)
